service/matching/matching_engine.go

```diff
@@ -864,6 +865,22 @@ func (e *matchingEngineImpl) QueryWorkflow(
 			return nil, serviceerror.NewInternal("unknown query completed type")
 		}
 	case <-ctx.Done():
+		// task timed out. log (optionally) and return the timeout error
+		ns, err := e.namespaceRegistry.GetNamespaceByID(namespaceID)
+		if err != nil {
+			e.logger.Error("Failed to get the namespace by ID",
+				tag.WorkflowNamespaceID(string(namespaceID)),
+				tag.Error(err))
+		} else {
+			sampleRate := e.config.QueryWorkflowTaskTimeoutLogRate(ns.Name().String())
+			if rand.Float64() < sampleRate {
+				e.logger.Info("Workflow Query Task timed out",
```

Comment:
We already have a `throttledLogger` available in most places.. are you sure we shouldn't use that?

Reply:
Discussed this with @yycptt . The concern is that the `throttledLogger` is shared. If we turn this on for a customer, it could spew a lot of log lines and consume full quota and other parts of matching engine logging may get throttled. Similarly if some other part triggers throttling, we may not see the log lines we expected to inspect.

Reply:
Sure, that's fair.

I've always been vaguely unhappy with the throttled logger anyway. It seems like what we want is a sampling+auto-adjusting rate limited logger: you construct it (at the top level) with a max rate, all calls pass a sample rate, and by default it samples at that rate. If the overall rate exceeds the max (using exponential average), it automatically adjusts all sample rates downward to stay below the overall max. Or maybe build in some fairness and have it bias toward keeping logs from less frequent call sites. Of course this is future work, this PR doesn't have to be blocked on that.